R/topic.R

```diff
@@ -0,0 +1,150 @@
+#' Gather topic from political discourses
+#'
+#' @param .data A data frame, "promises" data frame, or text vector.
+#' For data frames, function will search for "text" variable.
+#' For promises data frames coded using `select_promises()`,
+#' function will search for "promises" variable.
+#' @param dictionary The dictionary of 20 major political topics from the
+#' Comparative Agendas Project (Jones et al., 2023) is used by default.
+#' Users can also declare a custom dictionary as a vector or a list.
+#' If users declare a vector, each element is treated as a independent topic.
+#' If users declare a list of subjects and related terms, function understands
+#' names as topic and words as terms.
+#' @import dplyr
+#' @examples
+#' \donttest{
+#' gather_topics(US_News_Conferences_1960_1980[1:5, 3])
+#' gather_topics(US_News_Conferences_1960_1980[1:5, 3],
+#'                dictionary = c("military", "development"))
+#' gather_topics(.data = US_News_Conferences_1960_1980[1:5, 3],
+#'                dictionary = list("military" = c("military", "gun", "war"),
+#'                                  "development" = c("development", "interest rate", "banks")))
+#' }
+#' @export
+gather_topics <- function(.data, dictionary = "CAP") {
+  Words <- NULL
+  # get text variable
+  if (inherits(.data, "promises")) {
+    text <- stats::na.omit(.clean_token(getElement(.data, "promises")))
+  } else if (inherits(.data, "data.frame")) {
+    text <- .clean_token(getElement(.data, "text"))
+  } else text <- .clean_token(.data)
+  # get dictionary
+  if (any(dictionary ==  "CAP")) {
+    dictionary <- CAP_topics %>%
+      dplyr::mutate(Words = stringr::str_replace_all(textstem::lemmatize_words(
+        stringr::str_squish(tolower(Words))), ", ", "|"))
+    subjects <- dictionary$Words
+    names(subjects) <- dictionary$Topic
+    } else if (is.list(dictionary)) {
+      subjects <- lapply(dictionary, function(x) paste0(x, collapse = "|"))
```

Comment:
I could imagine some lists would kick an error here. Do you need to check that the list is a dictionary somehow?